engine/consensus/approvals/core.go

```diff
@@ -0,0 +1,392 @@
+		}
+	}
+
+	// in case block is not finalized we will create collector and start processing approvals
+	// no checks for orphans can be made at this point
+	// we expect that assignment collector will cleanup orphan IRs whenever new finalized block is processed
+
+	collector, newIncorporatedResult, err := c.getOrCreateCollector(result.Result)
+	if err != nil {
+		return fmt.Errorf("could not process incorporated result, cannot create collector: %w", err)
+	}
+
+	err = collector.ProcessIncorporatedResult(result)
+	if err != nil {
+		return fmt.Errorf("could not process incorporated result: %w", err)
+	}
+
+	// process pending approvals only if it's a new collector
+	// pending approvals are those we haven't received its result yet,
+	// once we received a result and created a new collector, we find the pending
+	// approvals for this result, and process them
+	// newIncorporatedResult should be true only for one goroutine even if multiple access this code at the same
+	// time, ensuring that processing of pending approvals happens once for particular assignment
+	if newIncorporatedResult {
+		err = c.processPendingApprovals(collector)
+		if err != nil {
+			return fmt.Errorf("could not process cached approvals:  %w", err)
+		}
+	}
+
+	return nil
+}
+
+// checkBlockOutdated performs a sanity check if block is outdated
+// Returns:
+// * engine.UnverifiableInputError - sentinel error in case we haven't discovered requested blockID
+// * engine.OutdatedInputError - sentinel error in case block is outdated
+// * exception in case of unknown internal error
+// * nil - block isn't sealed
+func (c *approvalProcessingCore) checkBlockOutdated(blockID flow.Identifier) error {
+	height, err := c.getCachedBlockHeight(blockID)
+	if err != nil {
+		return err
+	}
+
+	// it's important to use atomic operation to make sure that we have correct ordering
+	lastSealedHeight := c.lastSealedHeight()
+	// drop approval, if it is for block whose height is lower or equal to already sealed height
+	if lastSealedHeight >= height {
+		return engine.NewOutdatedInputErrorf("requested processing for already sealed block height")
+	}
+
+	return nil
+}
+
+func (c *approvalProcessingCore) ProcessApproval(approval *flow.ResultApproval) error {
+	err := c.checkBlockOutdated(approval.Body.BlockID)
+	if err != nil {
+		return fmt.Errorf("won't process approval for oudated block (%x): %w", approval.Body.BlockID, err)
+	}
+
+	if collector := c.getCollector(approval.Body.ExecutionResultID); collector != nil {
+		// if there is a collector it means that we have received execution result and we are ready
+		// to process approvals
+		err = collector.ProcessAssignment(approval)
+		if err != nil {
+			return fmt.Errorf("could not process assignment: %w", err)
+		}
+	} else {
+		// in case we haven't received execution result, cache it and process later.
+		c.approvalsCache.Put(approval)
```

Comment:
There is another important edge case here:

We learned this approval is for a unsealed block after `checkBlockOutdated` is called, but before `getCollector` is called, we received a new block that turns the block of the approval into a sealed block, (this is possible because `OnFinalizedBlock` is called in a different thread), which triggers `eraseCollectors ` to remove the collection, and `getCollector` returns nil. In this case, we still thought the block is "unsealed" and caches the approval since the result doesn't exist, but actually the block has been sealed, we just cached a sealed approval, which will stay in the cache for a long time.

The root cause is that the result of whether a block is sealed (or finalized) changes over time. In an event-driven model, we need to take these "outdated" event into consideration.

So here, I would suggest:
- we need to point out this edge case explicitly in comments
- Since whether a block is sealed is a changing event, and the logic that changes it lives in a different thread that is out of our control, we need to deal with the data consistency.
- To keep the data consistent, we could either:
    - make sure the state variable last sealed height always updated along with updates to `collectors` and `approvalCache` in order to ensure the sealed approvals and collectors have been removed, which I think will introduce a lot of locks and hurt the performance.
    - or to allow some inconsistency and ensure that even if these "outdated" approvals gets stored in the approval cache, they can be quickly or eventually removed to resume consistency and prevent memory leak.

Reply:
That is why I use lru cache storing it in cache won't actually hurt and will be ejected when new approvals discovered, I don't really think that we need to address it in some other explicit way.

Reply:
what's the main reason to use LRU in approval cache?

Is it to ensure stale approvals are ejected eventually.

Reply:
Yes, that's the main reasoning instead of using map or whatever + I am in process of adding index as you have suggested in your [comment](https://github.com/onflow/flow-go/pull/632#discussion_r620728013)